contrib/initramfs/hooks/zfsunlock.in

```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
+
+PREREQ="dropbear"
+
+prereqs() {
+    echo "$PREREQ"
+}
+
+case "$1" in
+    prereqs)
+        prereqs
+        exit 0
+    ;;
+esac
+
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+  cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
+#!/bin/sh
+if [ -e /run/zfsunlockcomplete ]; then
+  exit 0
+fi
+echo -n "Waiting for zfs init script to complete, press Ctrl-C to exit into shell ..."
+while [ ! -e /run/zfspassinpipe ] && [ ! -e /run/zfsunlockcomplete ]; do
+  sleep 1;
+  echo -n "."
+done
+if [ -e /run/zfsunlockcomplete ]; then
+  exit 0
+fi
+echo
+echo "Unlocking encrypted ZFS pools"
+echo "Enter correct password or press Ctrl-C to exit into shell"
+echo "After 3 incorrect passwords system will reboot"
+echo "Run zfsunlock command from shell to enter the password again"
+echo
+zfs_out_pipe=/run/zfspassinpipe
+zfs_in_pipe=/run/zfspassoutpipe
+zfs_fs_name=
+while read zfs_fs_name <$zfs_in_pipe; do
+  if [ $zfs_fs_name == "==mountfscompleted==" ]; then
+    break
+  fi
+  systemd-ask-password "Enter password for $zfs_fs_name:" > $zfs_out_pipe
+  read zfs_pass_status <$zfs_in_pipe
+  if [ $zfs_pass_status == "ok" ]; then
+    echo -e "password for $zfs_fs_name accepted\n"
+  elif [ $zfs_pass_status == "toomanyfailedattempts" ]; then
+    echo "too many password failures for $zfs_fs_name, rebooting"
+    reboot
+  else
+    echo "Wrong password, try again"
+  fi
+done
+echo -e "resuming boot sequence\nyou will be disconnected now, please reconnect in a while ..."
+EOF
+  chmod 0755 "$DESTDIR"/bin/zfsunlock
+  echo "zfsunlock" >> "$DESTDIR"/etc/profile
```

Comment:
Appending to /etc/profile feels completely wrong and almost surely needs to be redesigned.

Reply:
Agree, but for the moment I've got no other ideas how to automatically wait for the password input immediately after SSH login. Any useful hints on this regard ?

Reply:
I can think of a few options:
* Do nothing. Require the user to run a command (like `zfsunlock`).
* If dropbear or the initramfs environment support MOTD functionality, output a message describing the command to run.
* Modify `/root/.ssh/authorized_keys` to use the SSH `command=` parameter to force running the unlock command (assuming dropbear supports this).
* Use something in /etc/profile.d (if supported in the initramfs environment) or /etc/profile to print a message. That's basically the same, but seems less "bad".

Reply:
Tried to use /etc/motd
Unfortunately, for reasons unknown to me MOTD is completely ignored while running initramfs.
Tested on Debian 10.
I"ve modified the script to completely silence any output if there is no encrrypted partition or unlocking already has completed. Would it suffice the requirements ?